### PR TITLE
feat: convert structured ray logs to plain text when reading.

### PR DESCRIPTION
**Pull request by BenjaminMummery**
[ORQSDK-677]

# The problem

Ray logs are a mix of plain text (as output by Ray) and structured logs (as output by us). 

# This PR's solution

When reading ray logs, we check for lines that are parseable as json (indicating that they're structured) and convert them into plain text. 

This PR is a bare-minimum solution to the problem to serve as a fallback if other, preferable approaches prove impractical. 

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

[ORQSDK-677]: https://zapatacomputing.atlassian.net/browse/ORQSDK-677?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 1]
```diff
@@ -4,15 +4,19 @@
...
 import os
+import re
 import time
 import typing as t
 from dataclasses import dataclass
 from pathlib import Path
 
 from . import _client
 
+LINE_FORMAT: str = "{log.timestamp}\t{log.level} {log.filename} -- {log.message.logs}\t[{log.message.id}]"  # noqa E501
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/_ray/_ray_logs.py:18]: **nitpick (non-blocking):** ‏We can look into this later, but this format looks a little odd to me.

```
2023-01-30 11:03:13,150 INFO task_executor.py:78 -- Task status [RUNNING]       [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,157 INFO _log_adapter.py:108 -- hello       [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,157 CRITICAL tasks.py:7 -- HELLO    [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,169 INFO task_executor.py:78 -- Task status [RUNNING]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag_node.locals._ray_remote]
2023-01-30 11:03:13,180 INFO task_executor.py:78 -- Task status [RUNNING]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag.locals.handle_data_aggregation_error]
2023-01-30 11:03:10,935 INFO workflow_executor.py:86 -- Workflow job [id=wf.wf.cd86d3f] started.
2023-01-30 11:03:13,159 INFO workflow_executor.py:284 -- Task status [SUCCESSFUL]       [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,171 INFO workflow_executor.py:284 -- Task status [SUCCESSFUL]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag_node.locals._ray_remote]
2023-01-30 11:03:13,183 INFO workflow_executor.py:284 -- Task status [SUCCESSFUL]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag.locals.handle_data_aggregation_error]
2023-01-30 11:03:13,187 INFO workflow_executor.py:122 -- Workflow 'wf.wf.cd86d3f' completes successfully.
2023-01-30 11:03:13,150 INFO task_executor.py:78 -- Task status [RUNNING]       [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,157 INFO _log_adapter.py:108 -- hello       [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,157 CRITICAL tasks.py:7 -- HELLO    [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,169 INFO task_executor.py:78 -- Task status [RUNNING]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag_node.locals._ray_remote]
2023-01-30 11:03:13,180 INFO task_executor.py:78 -- Task status [RUNNING]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag.locals.handle_data_aggregation_error]
2023-01-30 11:03:10,935 INFO workflow_executor.py:86 -- Workflow job [id=wf.wf.cd86d3f] started.
2023-01-30 11:03:13,159 INFO workflow_executor.py:284 -- Task status [SUCCESSFUL]       [wf.wf.cd86d3f@invocation-0-task-a]
2023-01-30 11:03:13,171 INFO workflow_executor.py:284 -- Task status [SUCCESSFUL]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag_node.locals._ray_remote]
2023-01-30 11:03:13,183 INFO workflow_executor.py:284 -- Task status [SUCCESSFUL]       [wf.wf.cd86d3f@orquestra.sdk._ray._dag._make_ray_dag.locals.handle_data_aggregation_error]
```

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 2]
```diff
@@ -55,6 +73,74 @@ def _update_log_filenames(self):
...
+
+            data = {
+                "timestamp": m.group("timestamp"),
+                "level": m.group("level"),
+                "filename": m.group("filename"),
+                "message": {
+                    "logs": m.group("message"),
+                    "run_id": "",
+                },
+            }
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_ray_logs.py:103]: **issue**: regex-based JSON parsing is a very fragile implementation. Any change in the JSON output, like a new field, or a simple swap of fields would break it. Did you consider using `json.loads()` with a `try:` + `catch (ValueError, TypeError):`?

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 3]
```diff
@@ -55,6 +73,74 @@ def _update_log_filenames(self):
...
+        """
+        try:
+            ts_level_filename, logs = record.split(" -- ")
+            timestamp, level_filename = ts_level_filename.split("\t")
+            level, filename = level_filename.split(" ", 1)
+            try:
+                log_message, id = logs.split("\t")
+            except ValueError:
+                log_message = logs
+                id = ""
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_ray_logs.py:133]: **suggestion**: please don't use empty strings in absence of a value.

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 4]
```diff
@@ -55,6 +73,74 @@ def _update_log_filenames(self):
                     )
                 )
 
+    def _load_structured_log_record(self, record: str):
+        try:
+            data = ast.literal_eval(record)
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/_ray/_ray_logs.py:78]: **suggestion:** ‏I'm assuming you've used `ast.literal_eval` because the logs may not be parsed by the JSON module.
Can you add a comment for this and also add a note on `ast.literal_eval`'s safety? For example, "literal_eval only accepts a subset of Python code and cannot be used to perform malicious actions"

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 5]
```diff
@@ -64,7 +150,15 @@ def _refine_log_line(self, line):
         elif (
             self.workflow_or_task_run_id is None
         ) or self.workflow_or_task_run_id in line:
-            return line
+            if line.startswith("{"):
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_ray_logs.py:153]: **thought**: the `json.loads()` approach above would also tell us if a line is a JSON or plain text. This is a situation where EAFP might be better than LBYL. [Ext resource on the topic](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/).

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 6]
```diff
@@ -55,6 +73,74 @@ def _update_log_filenames(self):
...
+            if m is None:
+                return
+
+            data = {
+                "timestamp": m.group("timestamp"),
+                "level": m.group("level"),
+                "filename": m.group("filename"),
+                "message": {
+                    "logs": m.group("message"),
+                    "run_id": "",
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_ray_logs.py:101]: suggestion: please don't use empty strings in absence of a value. If there are cases where there's no `run_id` value we should use a different model class, or use `None`s.

File: src/orquestra/sdk/_ray/_ray_logs.py [hunk 7]
```diff
@@ -64,7 +150,15 @@ def _refine_log_line(self, line):
...
-            return line
+            if line.startswith("{"):
+                log = self._load_structured_log_record(line)
+            else:
+                log = self._load_unstructured_log_record(line)
+
+            if log is None:
+                return
+
+            return LINE_FORMAT.format(log=log).replace("[]", "").strip()
```
Comment (alexjuda) [inline src/orquestra/sdk/_ray/_ray_logs.py:161]: **question**: `.format()` followed by a `.replace("[]", "")`. Seems redundant. Can the format string be changed so get the right output in one step?

**Review by alexjuda — CHANGES_REQUESTED** [review]
Highlighted few issues. I know this PR serves as baseline, but just a few changes might make it much more robust.

General suggestion: could you add unit tests for this functionality? The only tests I could find were `tests/runtime/ray/test_integration.py::TestDirectRayReader` which is an integration test that takes a long time to run, and might not catch all the edge cases.

File: src/orquestra/sdk/_ray/_ray_logs.py (hunk 4 above)
Comment (alexjuda) [reply src/orquestra/sdk/_ray/_ray_logs.py:78]: Good catch, forgot to comment that. 

Parsing logs via `.literal_eval()` couples our logs format with Python syntax. We shouldn't do that unless there's a very good reason to.

**Comment by alexjuda** [comment]
@jamesclark-Zapata I'm gonna make a separate PR
